# Assignment 2

In this assigment, we will work with the *Forest Fire* data set. Please download the data from the [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/162/forest+fires). Extract the data files into the subdirectory: `../data/fires/` (relative to `./05_src/`).

## Objective

+ The model objective is to predict the area affected by forest fires given the features set. 
+ The objective of this exercise is to assess your ability to construct and evaluate model pipelines.
+ Please note: the instructions are not meant to be 100% prescriptive, but instead they are a set of minimum requirements. If you find predictive performance gains by applying additional steps, by all means show them. 

## Variable Description

From the description file contained in the archive (`forestfires.names`), we obtain the following variable descriptions:

1. X - x-axis spatial coordinate within the Montesinho park map: 1 to 9
2. Y - y-axis spatial coordinate within the Montesinho park map: 2 to 9
3. month - month of the year: "jan" to "dec" 
4. day - day of the week: "mon" to "sun"
5. FFMC - FFMC index from the FWI system: 18.7 to 96.20
6. DMC - DMC index from the FWI system: 1.1 to 291.3 
7. DC - DC index from the FWI system: 7.9 to 860.6 
8. ISI - ISI index from the FWI system: 0.0 to 56.10
9. temp - temperature in Celsius degrees: 2.2 to 33.30
10. RH - relative humidity in %: 15.0 to 100
11. wind - wind speed in km/h: 0.40 to 9.40 
12. rain - outside rain in mm/m2 : 0.0 to 6.4 
13. area - the burned area of the forest (in ha): 0.00 to 1090.84 









### Specific Tasks

+ Construct four model pipelines, out of combinations of the following components:

    + Preprocessors:

        - A simple processor that only scales numeric variables and recodes categorical variables.
        - A transformation preprocessor that scales numeric variables and applies a non-linear transformation.
    
    + Regressor:

        - A baseline regressor, which could be a [K-nearest neighbours model]() or a linear model like [Lasso](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Lasso.html) or [Ridge Regressors](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.ridge_regression.html).
        - An advanced regressor of your choice (e.g., Bagging, Boosting, SVR, etc.). TIP: select a tree-based method such that it does not take too long to run SHAP further below. 

+ Evaluate tune and evaluate each of the four model pipelines. 

    - Select a [performance metric](https://scikit-learn.org/stable/modules/linear_model.html) out of the following options: explained variance, max error, root mean squared error (RMSE), mean absolute error (MAE), r-squared.
    - *TIPS*: 
    
        * Out of the suggested metrics above, [some are correlation metrics, but this is a prediction problem](https://www.tmwr.org/performance#performance). Choose wisely (and don't choose the incorrect options.) 

+ Select the best-performing model and explain its predictions.

    - Provide local explanations.
    - Obtain global explanations and recommend a variable selection strategy.

+ Export your model as a pickle file.


You can work on the Jupyter notebook, as this experiment is fairly short (no need to use sacred). 

# Load the data

Place the files in the ../../05_src/data/fires/ directory and load the appropriate file. 

In [85]:
# Load the libraries as required.
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import Lasso
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import FunctionTransformer
from sklearn.preprocessing import PowerTransformer
import numpy as np

In [86]:
# Load data
columns = [
    'coord_x', 'coord_y', 'month', 'day', 'ffmc', 'dmc', 'dc', 'isi', 'temp', 'rh', 'wind', 'rain', 'area' 
]
fires_dt = (pd.read_csv('../../05_src/data/fires/forestfires.csv', header = 0, names = columns))
fires_dt.info()
# print(fires_dt.describe())
print("\n")
fires_dt.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 517 entries, 0 to 516
Data columns (total 13 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   coord_x  517 non-null    int64  
 1   coord_y  517 non-null    int64  
 2   month    517 non-null    object 
 3   day      517 non-null    object 
 4   ffmc     517 non-null    float64
 5   dmc      517 non-null    float64
 6   dc       517 non-null    float64
 7   isi      517 non-null    float64
 8   temp     517 non-null    float64
 9   rh       517 non-null    int64  
 10  wind     517 non-null    float64
 11  rain     517 non-null    float64
 12  area     517 non-null    float64
dtypes: float64(8), int64(3), object(2)
memory usage: 52.6+ KB




coord_x  coord_y month  day  ffmc   dmc     dc  isi  temp  rh  wind  rain  \
0        7        5   mar  fri  86.2  26.2   94.3  5.1   8.2  51   6.7   0.0   
1        7        4   oct  tue  90.6  35.4  669.1  6.7  18.0  33   0.9   0.0   
2        7        4   oct  sat  90.6  43.7  686.9  6.7  14.6  33   1.3   0.0   
3        8        6   mar  fri  91.7  33.3   77.5  9.0   8.3  97   4.0   0.2   
4        8        6   mar  sun  89.3  51.3  102.2  9.6  11.4  99   1.8   0.0   

   area  
0   0.0  
1   0.0  
2   0.0  
3   0.0  
4   0.0

# Get X and Y

Create the features data frame and target data.

In [87]:
# Prepare the data
X = fires_dt.drop('area', axis=1)
print(X.shape)

(517, 12)


In [88]:
# Define the target variable
y = fires_dt['area']
print(y.shape)

(517,)


In [89]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define numerical and categorical features
numerical_features = ['coord_x', 'coord_y', 'ffmc', 'dmc', 'dc', 'isi', 'temp', 'rh', 'wind', 'rain']
categorical_features = ['month', 'day']

# Display data split information
print("Data Split Summary:")
print(f"Training set shape: {X_train.shape}")
print(f"Test set shape: {X_test.shape}")
print(f"Numerical features: {numerical_features}")
print(f"Categorical features: {categorical_features}")

# Define a function for non-linear transformation (log transformation for positive values)
def log_transform(x):
    return np.log1p(x)

Data Split Summary:
Training set shape: (413, 12)
Test set shape: (104, 12)
Numerical features: ['coord_x', 'coord_y', 'ffmc', 'dmc', 'dc', 'isi', 'temp', 'rh', 'wind', 'rain']
Categorical features: ['month', 'day']


# Preprocessing

Create two [Column Transformers](https://scikit-learn.org/stable/modules/generated/sklearn.compose.ColumnTransformer.html), called preproc1 and preproc2, with the following guidelines:

- Numerical variables

    * (Preproc 1 and 2) Scaling: use a scaling method of your choice (Standard, Robust, Min-Max). 
    * Preproc 2 only: 
        
        + Choose a transformation for any of your input variables (or several of them). Evaluate if this transformation is convenient.
        + The choice of scaler is up to you.

- Categorical variables: 
    
    * (Preproc 1 and 2) Apply [one-hot encoding](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html) where appropriate.


+ The only difference between preproc1 and preproc2 is the non-linear transformation of the numerical variables.
    


### Preproc 1

Create preproc1 below.

+ Numeric: scaled variables, no other transforms.
+ Categorical: one-hot encoding.

In [90]:
# Create preproc1 - Numeric: scaled variables, no other transforms and Categorical: one-hot encoding.
preproc1 = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

### Preproc 2

Create preproc1 below.

+ Numeric: scaled variables, non-linear transformation to one or more variables.
+ Categorical: one-hot encoding.

In [91]:
# Create preproc2: Scaling, non-linear transformation, and one-hot encoding
preproc2 = ColumnTransformer(
    transformers=[
        ('num', Pipeline(steps=[
            ('scaler', StandardScaler()),
            ('transformer', PowerTransformer(method='yeo-johnson'))
        ]), numerical_features),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_features)
    ])

## Model Pipeline


Create a [model pipeline](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html): 

+ Add a step labelled `preprocessing` and assign the Column Transformer from the previous section.
+ Add a step labelled `regressor` and assign a regression model to it. 

## Regressor

+ Use a regression model to perform a prediction. 

    - Choose a baseline regressor, tune it (if necessary) using grid search, and evaluate it using cross-validation.
    - Choose a more advance regressor, tune it (if necessary) using grid search, and evaluate it using cross-validation.
    - Both model choices are up to you, feel free to experiment.

In [92]:
# Define the baseline and advanced models
baseline_model = Lasso()
advanced_model = RandomForestRegressor(random_state=42)

In [93]:
# Pipeline A = preproc1 + baseline
pipelineA = Pipeline(steps=[('preprocessing', preproc1),
                            ('regressor', baseline_model)])

In [94]:
# Pipeline B = preproc1 + advanced_model
pipelineB = Pipeline(steps=[('preprocessing', preproc1),
                            ('regressor', advanced_model)])

In [95]:
# Pipeline C = preproc2 + baseline_model
pipelineC = Pipeline(steps=[('preprocessing', preproc2),
                            ('regressor', baseline_model)])

In [96]:
# Pipeline D = preproc2 + advanced model
pipelineD = Pipeline(steps=[('preprocessing', preproc2),
                            ('regressor', advanced_model)])
    

# Tune Hyperparams

+ Perform GridSearch on each of the four pipelines. 
+ Tune at least one hyperparameter per pipeline.
+ Experiment with at least four value combinations per pipeline.

In [97]:
# Define parameter grids for GridSearchCV
param_gridA = {
    'regressor__alpha': [0.1, 1, 10, 100]
}

param_gridB = {
    'regressor__n_estimators': [50, 100, 200]
}

param_gridC = {
    'regressor__alpha': [0.1, 1, 10, 100]
}

param_gridD = {
    'regressor__n_estimators': [50, 100, 200]
}

In [98]:
# Create GridSearchCV objects for pipeline A
grid_searchA = GridSearchCV(pipelineA, param_gridA, cv=5, scoring='neg_mean_squared_error')

In [99]:
# Create GridSearchCV objects for pipeline B
grid_searchB = GridSearchCV(pipelineB, param_gridB, cv=5, scoring='neg_mean_squared_error')

In [100]:
# Create GridSearchCV objects for pipeline C
grid_searchC = GridSearchCV(pipelineC, param_gridC, cv=5, scoring='neg_mean_squared_error')

In [101]:
# Create GridSearchCV objects for pipeline D
grid_searchD = GridSearchCV(pipelineD, param_gridD, cv=5, scoring='neg_mean_squared_error')

In [102]:
# Fit the models
grid_searchA.fit(X_train, y_train)
grid_searchB.fit(X_train, y_train)
grid_searchC.fit(X_train, y_train)
grid_searchD.fit(X_train, y_train)

GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessing',
                                        ColumnTransformer(transformers=[('num',
                                                                         Pipeline(steps=[('scaler',
                                                                                          StandardScaler()),
                                                                                         ('transformer',
                                                                                          PowerTransformer())]),
                                                                         ['coord_x',
                                                                          'coord_y',
                                                                          'ffmc',
                                                                          'dmc',
                                                                          'dc',
                                                                          'isi',
                                                                          'temp',
                                                                          'rh',
                                                                          'wind',
                                                                          'rain']),
                                                                        ('cat',
                                                                         OneHotEncoder(handle_unknown='ignore'),
                                                                         ['month',
                                                                          'day'])])),
                                       ('regressor',
                                        RandomForestRegressor(random_state=42))]),
             param_grid={'regressor__n_estimators': [50, 100, 200]},
             scoring='neg_mean_squared_error')

In [103]:
# Model evaluation results
model_results = {
    'Pipeline A': {
        'Best Parameters': grid_searchA.best_params_,
        'RMSE': np.sqrt(-grid_searchA.best_score_)
    },
    'Pipeline B': {
        'Best Parameters': grid_searchB.best_params_,
        'RMSE': np.sqrt(-grid_searchB.best_score_)
    },
    'Pipeline C': {
        'Best Parameters': grid_searchC.best_params_,
        'RMSE': np.sqrt(-grid_searchC.best_score_)
    },
    'Pipeline D': {
        'Best Parameters': grid_searchD.best_params_,
        'RMSE': np.sqrt(-grid_searchD.best_score_)
    }
}


In [104]:
# Display the results
# results

print("Model Evaluation Results:")
for model_name, metrics in model_results.items():
    print(f"\n{model_name}:")
    print(f"*" * 20)
    print(f"Best Parameters: {metrics['Best Parameters']}")
    print(f"RMSE: {metrics['RMSE']:.4f}")

Model Evaluation Results:

Pipeline A:
********************
Best Parameters: {'regressor__alpha': 1}
RMSE: 45.7324

Pipeline B:
********************
Best Parameters: {'regressor__n_estimators': 100}
RMSE: 52.7634

Pipeline C:
********************
Best Parameters: {'regressor__alpha': 1}
RMSE: 45.6295

Pipeline D:
********************
Best Parameters: {'regressor__n_estimators': 100}
RMSE: 52.1451


# Evaluate

+ Which model has the best performance?

Based on the model evaluation result, the pipeline C seems to be best performing model in this case because of its evaluation matric RMSE value of 45.63 which is lowest among all pipelines

# Export

+ Save the best performing model to a pickle file.

In [105]:
# Save the best performing model to a pickle file
import pickle

best_model = grid_searchC.best_estimator_
with open('best_forest_fire_model.pkl', 'wb') as file:
    pickle.dump(best_model, file)
print("\nBest model saved to 'best_forest_fire_model.pkl'.")


Best model saved to 'best_forest_fire_model.pkl'.


# Explain

+ Use SHAP values to explain the following only for the best-performing model:

    - Select an observation in your test set and explain which are the most important features that explain that observation's specific prediction.

    - In general, across the complete training set, which features are the most and least important.

+ If you were to remove features from the model, which ones would you remove? Why? How would you test that these features are actually enhancing model performance?

*(Answer here.)*

## Criteria

The [rubric](./assignment_2_rubric_clean.xlsx) contains the criteria for assessment.

## Submission Information

🚨 **Please review our [Assignment Submission Guide](https://github.com/UofT-DSI/onboarding/blob/main/onboarding_documents/submissions.md)** 🚨 for detailed instructions on how to format, branch, and submit your work. Following these guidelines is crucial for your submissions to be evaluated correctly.

### Submission Parameters:
* Submission Due Date: `HH:MM AM/PM - DD/MM/YYYY`
* The branch name for your repo should be: `assignment-2`
* What to submit for this assignment:
    * This Jupyter Notebook (assignment_2.ipynb) should be populated and should be the only change in your pull request.
* What the pull request link should look like for this assignment: `https://github.com/<your_github_username>/production/pull/<pr_id>`
    * Open a private window in your browser. Copy and paste the link to your pull request into the address bar. Make sure you can see your pull request properly. This helps the technical facilitator and learning support staff review your submission easily.

Checklist:
- [ ] Created a branch with the correct naming convention.
- [ ] Ensured that the repository is public.
- [ ] Reviewed the PR description guidelines and adhered to them.
- [ ] Verify that the link is accessible in a private browser window.

If you encounter any difficulties or have questions, please don't hesitate to reach out to our team via our Slack at the `help` channel. Our Technical Facilitators and Learning Support staff are here to help you navigate any challenges.

# Reference

Cortez,Paulo and Morais,Anbal. (2008). Forest Fires. UCI Machine Learning Repository. https://doi.org/10.24432/C5D88D.